src/tree/updater_gpu_hist.cu

```diff
@@ -1001,30 +851,27 @@ struct DeviceShard {
     dh::safe_cuda(cudaMemcpy(d_nodes.data(), p_tree->GetNodes().data(),
                              d_nodes.size() * sizeof(RegTree::Node),
                              cudaMemcpyHostToDevice));
-    auto d_position = position.Current();
-    const auto d_ridx = ridx.Current();
-    auto d_matrix = this->ellpack_matrix;
-    dh::LaunchN(device_id, position.Size(), [=] __device__(size_t idx) {
-      auto position = d_position[idx];
-      auto node = d_nodes[position];
-      bst_uint ridx = d_ridx[idx];
-
-      while (!node.IsLeaf()) {
-        bst_float element = d_matrix.GetElement(ridx, node.SplitIndex());
-        // Missing value
-        if (isnan(element)) {
-          position = node.DefaultChild();
-        } else {
-          if (element <= node.SplitCond()) {
-            position = node.LeftChild();
-          } else {
-            position = node.RightChild();
+    auto d_matrix = ellpack_matrix;
+    row_partitioner->FinalisePosition(
```

Comment:
Interesting,  where do you plan to resue it?

Reply:
I don't necessarily plan to reuse it, but the benefits of extracting into components are still very nice. Testing and refactoring is much easier.